contrib/assets_tutorial/assets_tutorial.py

```diff
@@ -3,111 +3,185 @@
-e1 = startelementsd(e1_datadir, e1conf)
-time.sleep(2)
+class Daemon():
+    """
+    A class for representing a bitcoind or elementsd node.
+
+    Wraps the process management, creation and deletion of datadirs, and
+    RPC connectivity, into a simple object that will clean itself up on
+    exit. The `cleanup_on_exit` parameter can be set to False to prevent
+    the node from deleting its datadir on restart (and can be set by
+    passing --no-cleanup to the main program).
+    """
+
+    def __init__(self, name, daemon, path, conf_path, cleanup_on_exit = True):
+        self.name = name
+        self.daemon = daemon
+        self.conf_path = path
+        self.path = path
+        self.cleanup_on_exit = cleanup_on_exit
+        self.conf_path = conf_path
+        self.datadir_path = None
+        self.proc = None
+        self.rpc = None
+
+        # Parse config
+        self.config = {}
+        with open(self.conf_path, encoding="utf8") as f:
+            for line in f:
+                if len(line) == 0 or line[0] == "#" or len(line.split("=")) != 2:
+                    continue
+                self.config[line.split("=")[0]] = line.split("=")[1].strip()
+
+    def shutdown(self):
+        if self.proc is not None:
+            self.proc.kill()
+            self.proc = None
+
+        if self.datadir_path is not None:
+            if self.cleanup_on_exit:
+                shutil.rmtree(self.datadir_path)
+            else:
+                print ("Leaving %s datadir at %s." % (self.name, self.datadir_path))
+
+    def start(self, ext_args = None):
+        self.shutdown()
+
+        if ext_args is None:
+            ext_args = []
+        # Create datadir and copy config into place
+        self.datadir_path = tempfile.mkdtemp()
+        shutil.copyfile(self.conf_path, self.datadir_path + '/' + self.daemon + '.conf')
+        print("%s datadir: %s" % (self.name, self.datadir_path))
+        # Start process
+        self.proc = subprocess.Popen([self.path, "-datadir=" + self.datadir_path] + ext_args, stdout=subprocess.PIPE)
+        self.rpc = AuthServiceProxy("http://" + self.config["rpcuser"] + ":" + self.config["rpcpassword"] + "@127.0.0.1:" + self.config["rpcport"])
+
+    def restart(self, ext_args = None):
+        self.start(ext_args)
+
+    def __del__(self):
+        self.shutdown()
+
+    def __getattr__(self, name):
+        """Dispatches any unrecognised messages to the RPC connection or a CLI instance."""
+        return self.rpc.__getattr__(name)
+
+    def __getitem__(self, key):
+        """Dispatches any keys to the underlying config file"""
+        return self.config[key]
+
+def sync_all(nodes, timeout_sec = 10):
+    totalWait = timeout_sec
+
+    stop_time = time.time() + timeout
+    while time.time() <= stop_time:
+        best_hash = [x.getbestblockhash() for x in nodes]
+        if best_hash.count(best_hash[0]) == len(nodes):
+            return
+    raise Exception("Nodes cannot sync blocks or mempool!")
```

Comment:
You're no longer actually checking the mempool so this message became misleading.

Reply:
Nevermind, fixed in later commit.